codegenerator/cli/templates/static/codegen/src/eventFetching/ChainManager.res

```diff
@@ -269,18 +220,112 @@ let createBatchInternal = (
         let shouldNotAddItem = isInReorgThreshold && onlyBelowReorgThreshold
         if !shouldNotAddItem {
           popItemOffQueue()
-          batch->Js.Array2.push(item)->ignore
+          items->Js.Array2.push(item)->ignore
+          mutProcessingMetricsByChainId->Js.Dict.set(
+            item.chain->ChainMap.Chain.toChainId->Int.toString,
+            {
+              batchSize: switch mutProcessingMetricsByChainId->Utils.Dict.dangerouslyGetNonOption(
+                item.chain->ChainMap.Chain.toChainId->Int.toString,
+              ) {
+              | Some(metrics) => metrics.batchSize + 1
+              | None => 1
+              },
+              targetBlockNumber: item.blockNumber,
+            },
+          )
           loop()
         }
       }
     }
   loop()
 
-  {val: batch, isInReorgThreshold: isInReorgThresholdRef.contents}
+  (items, isInReorgThresholdRef.contents)
+}
+
+// Use a global pointer to spread the processing across chains
+let nextChainIdx = ref(0)
+
+let createUnorderedBatch = (
+  ~maxBatchSize,
+  ~fetchStatesMap: ChainMap.t<fetchStateWithData>,
+  ~onlyBelowReorgThreshold,
+  ~mutProcessingMetricsByChainId: dict<processingChainMetrics>,
+) => {
+  let isInReorgThresholdRef = ref(false)
+  let items = []
+
+  let chains = fetchStatesMap->ChainMap.keys
+  let unprocessedChains = ref(chains->Array.length) // Prevent entering the same chain twice
+  let batchSize = ref(0) // Faster than Array.length
+
+  // Accumulate items for all actively indexing chains
+  // the way to group as many items from a single chain as possible
+  // This way the loaders optimisations will hit more often
+  // Also, keep the nextChainIdx global, so we start with a new chain every batch
+  while batchSize.contents < maxBatchSize && unprocessedChains.contents > 0 {
+    let chainIdx = nextChainIdx.contents
+    switch chains->Array.get(chainIdx) {
+    | None => nextChainIdx := 0
+    | Some(chain) => {
+        let {fetchState, currentBlockHeight, highestBlockBelowThreshold} =
+          fetchStatesMap->ChainMap.get(chain)
+
+        // If the fetch state has reached the end block we don't need to consider it
+        if fetchState->FetchState.isActivelyIndexing {
+          let batchSizeBeforeTheChain = batchSize.contents
+
+          let rec loop = () =>
+            if batchSize.contents < maxBatchSize {
+              let earliestEvent = fetchState->FetchState.getEarliestEvent
+              let isInReorgThreshold =
+                earliestEvent->FetchState.queueItemIsInReorgThreshold(
+                  ~currentBlockHeight,
+                  ~highestBlockBelowThreshold,
+                )
+
+              isInReorgThresholdRef := isInReorgThresholdRef.contents || isInReorgThreshold
+
+              switch earliestEvent {
+              | NoItem(_) => ()
+              | Item({item, popItemOffQueue}) =>
+                //For dynamic contract pre registration, allow creating a batch up to the reorg threshold
```

Comment:
Think its worth mentioning so this behaviour doesn't get refactored out. (pre registration doesn't need to be considered anymore but the comment below is still important)
```suggestion
                // To ensure history saving only starts when all chains have reached their reorg threshold
                // otherwise one chain could sync very fast and force other chains to do historical sync 
                // while saving history which is much slower
```